internal/users/manager.go

```diff
@@ -292,33 +383,254 @@ func (m *Manager) AllShadows() ([]ShadowEntry, error) {
 	return shadowEntries, err
 }
 
-// GenerateUID deterministically generates an ID between from the given string, ignoring case,
-// in the range [UIDMin, UIDMax]. The generated ID is *not* guaranteed to be unique.
-func (m *Manager) GenerateUID(str string) uint32 {
-	return generateID(str, m.config.UIDMin, m.config.UIDMax)
+// RegisterUserPreAuth registers a temporary user with the given name and a unique UID in our NSS handler (in memory,
+// not in the database). It returns the generated UID or an error if the user could not be registered.
+func (m *Manager) RegisterUserPreAuth(name string) (uint32, error) {
+	m.temporaryEntriesMu.Lock()
+	defer m.temporaryEntriesMu.Unlock()
+
+	if m.numPreAuthUsers >= maxPreAuthUsers {
+		return 0, fmt.Errorf("maximum number of pre-auth users reached, login for new users via SSH is disabled, restart authd to enable it")
+	}
+
+	uid, _, err := m.registerUser(name, true)
+	if err != nil {
+		return 0, fmt.Errorf("could not register user %q: %w", name, err)
+	}
+
+	m.numPreAuthUsers++
+
+	return uid, nil
 }
 
-// GenerateGID deterministically generates an ID between from the given string, ignoring case,
-// in the range [GIDMin, GIDMax]. The generated ID is *not* guaranteed to be unique.
-func (m *Manager) GenerateGID(str string) uint32 {
-	return generateID(str, m.config.GIDMin, m.config.GIDMax)
+// registerUser registers a temporary user with the given name and a unique UID in our NSS handler (in memory, not in
+// the database).
+//
+// The caller must lock m.temporaryEntriesMu for writing before calling this function.
```

Comment:
maybe we should call the function `registerUserUnlocked` (and so for similar functions) to make it clearer?

Reply:
I don't find `registerUserUnlocked` easy to understand. Since it's an internal function which I don't expect being reused anywhere else than the two cases where it's currently used, I wouldn't bother making making the mutex requirement part of the function name.